### DISCUSSION TITLE
Suggested way of bundling model with extra artifacts

### DISCUSSION THREAD

**Discussion by MrMaks** [mlflow/mlflow#7537] [Q&A]
A common use case in both batch scoring/training pipelines and online inference endpoints is the need to bundle extra artifacts together with the model. Be it custom post-scoring logic which may be dynamic, preprocessing artifacts (think sklearn Scaler, Imputer) or for NLP use cases vocabulary files. While pytorch flavor allows that right now by extra_files parameter in log_model (https://github.com/mlflow/mlflow/pull/3436) this is not the case for other flavors. Thus far I see the only way of covering that generically is by wrapping every flavor model into pyfunc flavor and pass additional entries in artifacts argument, making them available in predict method via context dict. 

I cannot help but think there surely is a correct way of handling bundled model and artifacts, what would that be?  

One option I see is to utilize per-run artifact logging and when loading a model to establish a link to its run to access its artifacts. This seems counterintuitive where the model artifact should encompass all necessary objects as the final product of the training run.

**Comment by JHibbard** [comment]
Hi MrMaks,

Could you open up a feature request in the MLflow repo with your examples of where this would be an advantage? I don't have a universal answer for this requirement, although some use cases may be covered by the MLflow Recipes feature with its transform and preprocessing steps.

Hope that helps!
Jim
